Был загружен датасет с помощью библиотеки pandas

In [ ]:
import pandas as pd
df = pd.read_csv("/content/credit_risk.csv")

Были выведены первые 20 строк с помощью метода head

In [ ]:
df.head(20)

Id   Age  Income      Home  Emp_length             Intent  Amount   Rate  \
0    0  22.0   59000      RENT       123.0           PERSONAL   35000  16.02   
1    1  21.0    9600       OWN         5.0          EDUCATION    1000  11.14   
2    2  25.0    9600  MORTGAGE         1.0            MEDICAL    5500  12.87   
3    3  23.0   65500      RENT         4.0            MEDICAL   35000  15.23   
4    4  24.0   54400      RENT         8.0            MEDICAL   35000  14.27   
5    5  21.0    9900       OWN         2.0            VENTURE    2500   7.14   
6    6  26.0   77100     RENET         8.0          EDUCATION   35000  12.42   
7    7  24.0   78956      RENT         5.0            MEDICAL   35000  11.11   
8    8  24.0   83000      RENT         8.0           PERSONAL   35000   8.90   
9    9  21.0   10000       OWN         6.0            VENTURE    1600  14.74   
10  10  22.0   85000      RENT         6.0            VENTURE   35000  10.37   
11  11  21.0   10000       OWN         2.0    HOMEIMPROVEMENT    4500   8.63   
12  12  23.0   95000      RENT         2.0            VENTURE   35000   7.90   
13  13  26.0  108160      RENT         4.0          EDUCATION   35000  18.39   
14  14  23.0  115000      RENT         2.0          EDUCATION   35000   7.90   
15  15  23.0  500000  MORTGAGE         7.0  DEBTCONSOLIDATION   30000  10.65   
16  16  23.0  120000      RENT         0.0          EDUCATION   35000   7.90   
17  17  23.0   92111      RENT         7.0            MEDICAL   35000  20.25   
18  18  23.0  113000      RENT         8.0  DEBTCONSOLIDATION   35000  18.25   
19  19  24.0   10800  MORTGAGE         8.0          EDUCATION    1750  10.99   

    Status  Percent_income Default  Cred_length  
0        1            0.59       Y            3  
1        0            0.10       N            2  
2        1            0.57       N            3  
3        1            0.53       N            2  
4        1            0.55       Y            4  
5        1            0.25       N            2  
6        1            0.45       N            3  
7        1            0.44       N            4  
8        1            0.42       N            2  
9        1            0.16       N            3  
10       1            0.41       N            4  
11       1            0.45       N            2  
12       1            0.37       N            2  
13       1            0.32       N            4  
14       0            0.30       N            4  
15       0            0.06       N            3  
16       0            0.29       N            4  
17       1            0.32       N            4  
18       1            0.31       N            4  
19       1            0.16       N            2

# Обзор каждого столбца датасета:
>**Возраст** (Age) - содержит информацию о возрасте людей, подавших заявку на кредит.тупами

> **Доход**(Income) - содержит информацию о ежемесячном доходе.


> **Статус домовладения** (Home) (Собственный, Ипотека, Аренда) - столбец
содержит информацию о статусе домовладения заявителя.


>**Стаж работы в годах**(Emp_length) - столбец содержит информацию о стаже работы в годах.


> **Цель кредита**(Intent) - столбец содержит информацию о цели получения
кредита.



> **Сумма кредита по заявке**(Amount) - столбец содержит информацию о
запрашиваемой сумме кредита.




> **Процентная ставка**(Rate) - столбец содержит информацию о процентной
ставке кредита.




> **Статус одобрения кредита**(Status) - столбец содержит информацию о статусе одобрения кредита(1, 0).





>**Сумма кредита в процентах от дохода**(Percent_income) - столбец показывает
соотношение суммы кредита к доходу заявителя в процентах.




> **Не допустил ли заявитель ранее дефолта по кредиту (да, нет)** (Default) -
столбец указывает допускал ли заявитель ранее дефолт по кредиту.



> **Длина кредитной истории заявителя** (Credit_length) - столбец содержит информацию о длине кредитной истории заявителя (в годах).




Проанализировав эти данные, банку будет проще принять решение о выдаче кредита заявителю.





С помощью метода info была выведена информация о данных. Благодаря этой информации, можно сделать вывод, что всего 652 строки, 12 столбцов, в столбце «стаж работы в годах» присутствует 9 строк с нулевыми данными, в столбце «процентная ставка» присутствует 66 строк с нулевыми данными, типы данных в таблице - float64, int64, object.

In [ ]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 652 entries, 0 to 651
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Id              652 non-null    int64  
 1   Age             652 non-null    float64
 2   Income          652 non-null    int64  
 3   Home            652 non-null    object 
 4   Emp_length      643 non-null    float64
 5   Intent          652 non-null    object 
 6   Amount          652 non-null    int64  
 7   Rate            586 non-null    float64
 8   Status          652 non-null    int64  
 9   Percent_income  652 non-null    float64
 10  Default         652 non-null    object 
 11  Cred_length     652 non-null    int64  
dtypes: float64(4), int64(5), object(3)
memory usage: 61.2+ KB


Был вызван метод describe, который позволяет оценить числовые столбцы таблицы. Он выводит: количество значений, среднее значение, стандартное отклонение значений, минимальное значение, первый квартиль, который означает, что 25% значений в столбце ниже максимального значения, медиана, которая означает, что 50% значений ниже максимального значения, и третий квартиль, который означает, что 75% значений ниже максимального значения.

In [ ]:
df.describe()


Id         Age         Income  Emp_length        Amount  \
count  652.000000  652.000000     652.000000  643.000000    652.000000   
mean   325.495399   24.285276   90008.855828    4.601866  18800.996933   
std    188.352535    7.841061   69530.312525    7.281616   9152.896250   
min      0.000000   21.000000    9600.000000    0.000000   1000.000000   
25%    162.750000   23.000000   44000.000000    2.000000  10000.000000   
50%    325.500000   24.000000   69998.000000    4.000000  21850.000000   
75%    488.250000   25.000000  128499.000000    7.000000  25000.000000   
max    649.000000  144.000000  500000.000000  123.000000  35000.000000   

             Rate      Status  Percent_income  Cred_length  
count  586.000000  652.000000      652.000000   652.000000  
mean    12.289471    0.601227        0.279831     3.012270  
std      3.272772    0.490022        0.150231     0.808843  
min      5.420000    0.000000        0.010000     2.000000  
25%     10.250000    0.000000        0.160000     2.000000  
50%     12.180000    1.000000        0.280000     3.000000  
75%     14.720000    1.000000        0.380000     4.000000  
max     21.210000    1.000000        0.830000     4.000000

По выведенным данным, можно сделать вывод, что средний возраст заявителя - 24 года, максимальный доход - 500000, минимальный стаж работы - 0 лет, средняя сумма кредита по заявке - 18800, минимальная средняя ставка - 5.4 процента, по среднему значению статуса заявки (0.6) можем сделать вывод, что количество одобренных кредитов немного превышает количество отказов, средняя сумма месячной выплаты кредита в процентах от дохода - 27 процентов, а кредитная история в среднем - 3 года.

С помощью метода columns были выведены названия столбцов. Названия не потребуют изменений.

In [ ]:
df.columns


Index(['Id', 'Age', 'Income', 'Home', 'Emp_length', 'Intent', 'Amount', 'Rate',
       'Status', 'Percent_income', 'Default', 'Cred_length'],
      dtype='object')

С помощью метода isna было выведено количество строк с пропусками.

In [ ]:
df.isna().sum()

Id                 0
Age                0
Income             0
Home               0
Emp_length         9
Intent             0
Amount             0
Rate              66
Status             0
Percent_income     0
Default            0
Cred_length        0
dtype: int64

 С помощью метода filna пропуски в столбце процентная ставка были заменены на 10 - среднее значение процентной ставки в банке, стаж работы в годах - на 0, что может повлиять на вероятность выдачи кредита, но, данное значение подходит, в случае, если заявитель не заполнил данную строку.

In [ ]:
df['Emp_length'] = df['Emp_length'].fillna(0)
df['Rate'] = df['Rate'].fillna(10)
df.isna().sum()

Id                0
Age               0
Income            0
Home              0
Emp_length        0
Intent            0
Amount            0
Rate              0
Status            0
Percent_income    0
Default           0
Cred_length       0
dtype: int64

С помощью метода duplicated была получена информация о количестве дубликатов (строк, которые полностью повторяют уже встречавшиеся в датасете строки)

In [ ]:
df.duplicated().sum()


2

Также были найдены продублированные строки. Дубликаты были найдены в столбце "Home"

In [ ]:
df[df.duplicated()]

Id   Age  Income  Home  Emp_length             Intent  Amount  Rate  \
650  649  23.0   58800  RENT         7.0  DEBTCONSOLIDATION   20000  7.66   
651  649  23.0   58800  RENT         7.0  DEBTCONSOLIDATION   20000  7.66   

     Status  Percent_income Default  Cred_length  
650       1            0.34       N            3  
651       1            0.34       N            3

С помощью метода drop_duplicates эти дубликаты были удалены.

In [ ]:
df = df.drop_duplicates().reset_index(drop=True)

Была проведена проверка с помощью вызова функции duplicated ещё раз.

In [ ]:
df.duplicated().sum()

0

С помощью метода unique были найдены неявные дубликаты в столбце «Статус домовладения» и исправлены неправильные и альтернативные значения

In [ ]:
df['Home'].unique()

array(['RENT', 'OWN', 'MORTGAGE', 'RENET', 'OTHER'], dtype=object)

In [ ]:
df['Home'] = df['Home'].replace('RENET','RENT')
df['Home'].unique()

array(['RENT', 'OWN', 'MORTGAGE', 'OTHER'], dtype=object)

Также были проверен столбец "Цель кредита"

In [ ]:
print(df['Intent'].unique())

['PERSONAL' 'EDUCATION' 'MEDICAL' 'VENTURE' 'HOMEIMPROVEMENT'
 'DEBTCONSOLIDATION']



Также был проверен столбец «Не допустил ли заявитель ранее дефолта по кредиту (да, нет)»

In [ ]:
df['Default'].unique()

array(['Y', 'N', 'No'], dtype=object)

В нём был выявлен и исправлен дубликат.

In [ ]:
df['Default'] = df['Default'].replace('No','N')
df['Default'].unique()

array(['Y', 'N'], dtype=object)

С помощью метода dtypes были проверены типы данных.

In [ ]:
df.dtypes

Id                  int64
Age               float64
Income              int64
Home               object
Emp_length        float64
Intent             object
Amount              int64
Rate              float64
Status              int64
Percent_income    float64
Default            object
Cred_length         int64
dtype: object

В столбце Age подходящий тип данных - int64, у нас же - float64. Необходимо изменить тип данных в столбце Age. Для этого вызовем метод astype.

In [ ]:
df['Age'] = df['Age'].astype(int)

После вызова метода необходимо проверить, изменился ли тип данных в изначальном датафрейме.

In [ ]:
df.dtypes

Id                  int64
Age                 int64
Income              int64
Home               object
Emp_length        float64
Intent             object
Amount              int64
Rate              float64
Status              int64
Percent_income    float64
Default            object
Cred_length         int64
dtype: object


# **Группировка**


Группировка - “employment_type”  и количество компаний по каждой локации “company_location”. Создать датафрейм. Переименовать столбец с количеством в “сount”. Отсортировать по возрастанию столбца “count”.

С помощью метода groupby() данные были сгруппированы по цели кредита и количеству статуса домовладения. Подсчёт был сделан с помощью count.

In [ ]:
df.groupby(['Intent','Home'])['Id'].count()

Intent             Home    
DEBTCONSOLIDATION  MORTGAGE    21
                   OWN          9
                   RENT        72
EDUCATION          MORTGAGE    34
                   OTHER        1
                   OWN         16
                   RENT        94
HOMEIMPROVEMENT    MORTGAGE    13
                   OWN         16
                   RENT        39
MEDICAL            MORTGAGE    22
                   OWN         15
                   RENT        70
PERSONAL           MORTGAGE    19
                   OTHER        1
                   OWN         22
                   RENT        69
VENTURE            MORTGAGE    27
                   OWN         20
                   RENT        70
Name: Id, dtype: int64

По данной группировке можно сделать вывод, что большинство заявителей ариндует жилье, самый большой процент заявителей берут кредит на обучение, наибольшее число заявителей с собственным жильем берут кредит на личные цели.

Группировка - “employment_type”  и кпо каждой локации “company_location”. Результат должен быть выведен в следующем формате (на скриншоте представлен фрагмент).

С помощью Groupby создаётся датафрейм, в котором с помощью метода size и reset_index переименовывается столбец подсчета в Count, с помощью sort.values значения сортируются, с помощью ascending = False сортируются по убыванию.

In [ ]:
group1 = df.groupby(['Intent','Home'])
answer_group = group1.size().reset_index(name='count')
answer_group = answer_group.sort_values(by = 'count', ascending = False)
answer_group

Intent      Home  count
6           EDUCATION      RENT     94
2   DEBTCONSOLIDATION      RENT     72
19            VENTURE      RENT     70
12            MEDICAL      RENT     70
16           PERSONAL      RENT     69
9     HOMEIMPROVEMENT      RENT     39
3           EDUCATION  MORTGAGE     34
17            VENTURE  MORTGAGE     27
15           PERSONAL       OWN     22
10            MEDICAL  MORTGAGE     22
0   DEBTCONSOLIDATION  MORTGAGE     21
18            VENTURE       OWN     20
13           PERSONAL  MORTGAGE     19
8     HOMEIMPROVEMENT       OWN     16
5           EDUCATION       OWN     16
11            MEDICAL       OWN     15
7     HOMEIMPROVEMENT  MORTGAGE     13
1   DEBTCONSOLIDATION       OWN      9
14           PERSONAL     OTHER      1
4           EDUCATION     OTHER      1

По данной группировке можно сделать вывод, что большинство заявителей арендует жилье, также, наибольшее число заявителей берут кредит на образование. Наименьшее количество заявителей указали тип домовладения "Другой", и цель кредита - "Обучение" и "Личные цели".

Сводная таблица (pivot_table) - средняя зарплата по годам (”work_year”). Отсортировать по убыванию зп. Округлить до трёх знаков после запятой. Переименовать столбец “salary” в “зарплата”.

С помощью метода pivot_table была создана сводная таблица, значение – Income (значения, которые нужно сгруппировать), индекс – Age (столбец, по которому данные группируются), arggfunc – функция, которая применяется к значениям, в моём случае функция, которая берет средние значения.

In [ ]:
pivot_3 = df.pivot_table(values = 'Income', index = 'Age', aggfunc = 'mean')
pivot_3 = pivot_3.sort_values(by = 'Income', ascending = False)
pivot_3['Income'] = pivot_3['Income'].round(2)
pivot_3

Income
Age           
144  225000.00
26   122793.74
25   106750.02
24    88684.03
23    85682.35
123   80004.00
22    64423.88
21    42238.58

По данной сводной таблице можно сделать вывод, что наибольшая сумма дохода у заявителей с возрастом 26, наименьшее - 21. Можно также заметить, что с ростом возраста вырастает и зарплата.

Сводная таблица (pivot_table) - медианная зарплата по годам - строки, и по “employment_type” - столбцы. Отсортировать по убыванию годов.

С помощью метода pivot_table была создана сводная таблица, значение – Income (значения, которые нужно сгруппировать), индекс – Home (столбец, по которому данные группируются), columns – Intent (столбец, по которому происходит группировка), arggfunc – функция, которая применяется к значениям, в моём случае функция, которая берет медианные значения.

In [ ]:

pivot_4 = df.pivot_table(values = 'Income', index = 'Home', columns = 'Intent', aggfunc = 'median')
pivot_4 = pivot_4.sort_values(by = 'Home', ascending = False)
pivot_4 = pivot_4.round(2)
pd.pivot_table(df, values = 'Income', index = 'Home', columns = 'Intent', fill_value = 0 )

Intent    DEBTCONSOLIDATION      EDUCATION  HOMEIMPROVEMENT        MEDICAL  \
Home                                                                         
MORTGAGE      128852.047619  141393.411765    144012.615385  117061.545455   
OTHER              0.000000   11389.000000         0.000000       0.000000   
OWN            14334.666667   63824.750000     72565.437500   39385.066667   
RENT           85176.986111   87780.276596     83998.538462   79992.628571   

Intent         PERSONAL        VENTURE  
Home                                    
MORTGAGE  102129.526316  145813.370370  
OTHER      19200.000000       0.000000  
OWN        69722.545455   44181.000000  
RENT       86012.550725   88363.528571

По данной сводной таблице можно сделать вывод, что наибольшая медианная зарплата у заявителей с типом домовладения "Mortage" и целью кредита "Venture", наименьшая же - тип домовладения "Own" и цель кредита - "Medical". Также можно заметить, что заявители с типом домовладения "Other" брали кредит только с целью "Education" и "Personal"

Необходимо отфильтровать возраст до 100, создать новый датасет из них, выбрать цель кредита - education, personal, medical. Создать сводную таблицу по этим данным с любыми столбцами, но values = rate, aggfunc = mean. Необходимо отфильтровать возраст до 100, создать новый датасет из них, выбрать цель кредита - education, personal, medical. Создать сводную таблицу по этим данным с любыми столбцами, но values = rate, aggfunc = mean. Создаём копию изначального датафрейма с помощью метода .copy(), с помощью метода .loc сортируем возраст, с помощью метода .isin() фильтруем цели кредита, после этого создаём сводную таблицу по новым данным.

In [ ]:
df2 = df.copy()
df2 = df.loc[df['Age'] <= 100]
Intents = ["MEDICAL", "EDUCATION", "PERSONAL"]
df2 = df2[df2['Intent'].isin(Intents)].reset_index()
pivot_dop = df2.pivot_table(index = 'Age', values = 'Rate', columns = 'Intent', aggfunc = 'mean')
pivot_dop

Intent  EDUCATION    MEDICAL   PERSONAL
Age                                    
21      11.461538  10.072000  12.983333
22      11.916562  12.525000  12.064286
23      11.431875  13.664783  11.366800
24      11.672381  12.281786  12.067000
25      11.426667  13.380000  12.334000
26      12.537619  13.046000  11.354706

По данной сводной таблице можно сделать вывод, что наибольшая процентная ставка по кредиту у заявителей с возрастом 25 и целью кредита "Medical", наименьшая же у заявителей с возрастом 21 и целью кредита "Medical", в среднем же, по цели кредита "Education" процентная ставка - 11%, по "Medical" - 13%, по "Personal" - 12%.

 Были проанализированы данные банка по заявкам на кредит. Данные содержат в себе информацию о типе домовладения заявителя, проценте кредита от зарплаты, стаже работы, и другие. По полученным данным можно сделать вывод о том, что на обучение было взято большинство кредитов, у двадцатишестилетних заявителей процентная ставка на образование самая высокая.
